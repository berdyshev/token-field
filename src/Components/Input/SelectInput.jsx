import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Checkbox, Input } from 'deskpro-components/lib/Components/Forms';
import Icon from 'deskpro-components/lib/Components/Icon';
import { List, ListElement, Scrollbar } from 'deskpro-components/lib/Components/Common';
import styles from 'styles/style.css';
import TokenInput from './TokenInput';

export default class SelectInput extends React.Component {
  static propTypes = {
    token: PropTypes.shape({
      type:  PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
    }).isRequired,
    dataSource: PropTypes.shape({
      getOptions:  PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
      findOptions: PropTypes.func,
    }).isRequired,
    isMultiple:          PropTypes.bool,
    className:           PropTypes.string,
    onChange:            PropTypes.func,
    selectPreviousToken: PropTypes.func.isRequired,
    selectNextToken:     PropTypes.func.isRequired,
    renderHeader:        PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.node,
    ]),
    renderItem:   PropTypes.func,
    renderFooter: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.node,
    ]),
    showSearch:            PropTypes.bool,
    selectionsTranslation: PropTypes.string,
    removeToken:           PropTypes.func.isRequired,
  };

  static defaultProps = {
    className:             '',
    onChange() {},
    isMultiple:            false,
    renderHeader:          null,
    renderItem:            null,
    renderFooter:          null,
    showSearch:            true,
    selectionsTranslation: 'selections'
  };

  static getIcon(option) {
    if (typeof option.icon === 'string') {
      return <Icon name={option.icon} />;
    }
    return null;
  }

  static getLabel(option) {
    if (option.label) {
      return option.label;
    }
    if (option.title) {
      return option.title;
    }
    if (option.name) {
      return option.name;
    }
    if (option.id) {
      return option.id;
    }
    return option.value;
  }

  constructor(props) {
    super(props);

    const options = this.props.dataSource.getOptions;

    this.state = {
      value:          props.token.value,
      selectedOption: options[0],
      options,
      filter:         '',
    };
  }

  componentWillUnmount() {
    window.document.removeEventListener('keydown', this.handleKeyDown);
  }

  onFocus = () => {
    if (this.props.showSearch) {
      this.searchInput.focus();
    }
    window.document.addEventListener('keydown', this.handleKeyDown);
  };

  onBlur = () => {
    window.document.removeEventListener('keydown', this.handleKeyDown);
  };

  focus = () => {
    this.tokenInput.focus();
  };

  handleChange = (option) => {
    const value = option.id || option.value;
    this.setState({
      value,
      filter: ''
    });
    this.props.onChange(value);
    this.tokenInput.disableEditMode();
  };

  handleKeyDown = (e) => {
    switch (e.key) {
      case 'ArrowDown':
      case 'ArrowUp': {
        const { options } = this.state;
        const index = options.findIndex(option => option === this.state.selectedOption);
        if (e.key === 'ArrowDown' && index < options.length - 1) {
          this.setState({
            selectedOption: options[index + 1]
          });
        }
        if (e.key === 'ArrowUp' && index > 0) {
          this.setState({
            selectedOption: options[index - 1]
          });
        }
        break;
      }
      case 'Escape':
        this.setState({
          value: this.props.token.value
        });
        this.tokenInput.disableEditMode();
        break;
      case 'Tab':
        if (e.shiftKey) {
          this.props.selectPreviousToken();
        } else {
          this.handleChange(this.state.selectedOption);
          this.props.selectNextToken();
        }
        this.tokenInput.disableEditMode();
        break;
      case 'Enter':
        this.handleChange(this.state.selectedOption);
        this.props.selectNextToken();
        break;
      default:
        return true;
    }
    e.preventDefault();
    return true;
  };

  handleChangeMultiple = (checked, value) => {
    const values = this.state.value;
    if (checked) {
      values.push(value);
    } else {
      const index = values.indexOf(value);
      values.splice(index, 1);
    }
    this.setState({
      value: values
    });
  };

  handleFilter = (filter) => {
    let { selectedOption } = this.state;
    const options = this.props.dataSource.getOptions
      .filter(option => filter === ''
        || SelectInput.getLabel(option).toLowerCase().indexOf(filter.toLowerCase()) !== -1);
    if (options.length) {
      if (!selectedOption || options.indexOf(selectedOption) === -1) {
        selectedOption = options[0];
      }
    } else {
      selectedOption = null;
    }
    this.setState({
      filter,
      options,
      selectedOption,
    });
  };

  renderInput = () => {
    const { isMultiple, showSearch } = this.props;
    return (
      <div className="dp-select">
        <div className="dp-select__content">
          <Scrollbar>
            <List className="dp-selectable-list">
              { showSearch ?
                <Input
                  name="search"
                  icon="search"
                  value={this.state.filter}
                  ref={(c) => { this.searchInput = c; }}
                  onChange={this.handleFilter}
                />
                : null }
              {this.renderHeader()}
              {isMultiple ? this.renderMultipleOptions() : this.renderOptions()}
              {this.renderFooter()}
            </List>
          </Scrollbar>
        </div>
      </div>
    );
  };

  renderValue = () => {
    const { getOptions, findOptions } = this.props.dataSource;
    let valueOption;
    let value;
    if (this.props.isMultiple) {
      if (this.state.value.length > 1) {
        return `${this.state.value.length} ${this.props.selectionsTranslation}`;
      }
      value = this.state.value[0];
    } else {
      value = this.state.value;
    }
    if (findOptions) {
      valueOption = findOptions(value);
    } else {
      valueOption = getOptions.find((option) => {
        if (option.id) {
          return option.id === value;
        }
        return option.value === value;
      });
    }
    if (!valueOption) {
      return '________';
    }
    return this.renderItem(valueOption);
  };

  renderHeader = () => {
    const { renderHeader } = this.props;
    if (typeof renderHeader === 'function') {
      return renderHeader();
    }
    return renderHeader;
  };

  renderOptions = () => {
    const { value, selectedOption, options } = this.state;
    return (
      options.map((option) => {
        const key = option.id || option.value;
        const currentValue = key === value ? styles['current-value'] : '';
        const selected = option === selectedOption ? styles.selected : '';
        return (
          <ListElement
            key={key}
            className={classNames(currentValue, selected)}
            onClick={() => this.handleChange(option)}
          >
            {this.renderItem(option)}
          </ListElement>
        );
      }));
  };

  renderItem = (option) => {
    if (this.props.renderItem) {
      return this.props.renderItem(option);
    }
    return (
      <span>
        {SelectInput.getIcon(option)}
        {SelectInput.getLabel(option)}
      </span>
    );
  };

  renderMultipleOptions() {
    const { getOptions } = this.props.dataSource;
    const { filter, value } = this.state;
    return (getOptions
      .filter(option => filter === ''
        || SelectInput.getLabel(option).toLowerCase().indexOf(filter.toLowerCase()) !== -1)
      .map((option) => {
        const key = option.id || option.value;
        const checked = value.indexOf(key) !== -1;
        return (
          <ListElement
            key={key}
          >
            <Checkbox checked={checked} value={key} onChange={this.handleChangeMultiple}>
              {this.renderItem(option)}
            </Checkbox>
          </ListElement>
        );
      }));
  }

  renderFooter = () => {
    const { renderFooter } = this.props;
    if (typeof renderFooter === 'function') {
      return renderFooter();
    }
    return renderFooter;
  };

  render() {
    const { token, className, removeToken } = this.props;
    return (
      <TokenInput
        ref={(c) => { this.tokenInput = c; }}
        className={className}
        type={token.type}
        renderInput={this.renderInput}
        renderValue={this.renderValue}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        removeToken={removeToken}
      />
    );
  }
}
