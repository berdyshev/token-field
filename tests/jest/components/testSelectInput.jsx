import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import SelectInput from 'Components/Input/SelectInput';
import noop from 'deskpro-components/lib/utils/noop';

let wrapper;
let token;

const options = [
  { label: 'Austria', value: 'AT' },
  { label: 'Belgium', value: 'BE' },
  { label: 'Bulgaria', value: 'BG' },
  { label: 'Croatia', value: 'HR' },
  { label: 'Cyprus', value: 'CY' },
  { label: 'Czech Republic', value: 'CZ' },
  { label: 'Denmark', value: 'DK' },
  { label: 'Estonia', value: 'EE' },
  { label: 'Finland', value: 'FI' },
  { label: 'France', value: 'FR' },
  { label: 'Germany', value: 'DE' },
  { label: 'Greece', value: 'GR' },
  { label: 'Hungary', value: 'HU' },
  { label: 'Ireland', value: 'IE' },
  { label: 'Italy', value: 'IT' },
  { label: 'Latvia', value: 'LV' },
  { label: 'Lithuania', value: 'LT' },
  { label: 'Luxembourg', value: 'LU' },
  { label: 'Malta', value: 'MT' },
  { label: 'Netherlands', value: 'NL' },
  { label: 'Poland', value: 'PL' },
  { label: 'Portugal', value: 'PT' },
  { label: 'Romania', value: 'RO' },
  { label: 'Slovakia', value: 'SK' },
  { label: 'Slovenia', value: 'SI' },
  { label: 'Spain', value: 'ES' },
  { label: 'Sweden', value: 'SE' },
  { label: 'United Kingdom', value: 'GB' }
];

it('+++capturing Snapshot of SelectInput', () => {
  const renderedValue = renderer.create(
    <SelectInput
      dataSource={{ getOptions: options }}
      token={token}
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});

beforeEach(() => {
  token = {
    type:  'country',
    value: 'GB'
  };
  wrapper = mount(
    <SelectInput
      dataSource={{ getOptions: options }}
      token={token}
      className="test"
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  );
});

it('should render the input label', () => {
  expect(wrapper.contains(<div className="dp-code label">{token.type}:</div>)).toEqual(true);
});

it('should render the input value', () => {
  const span = wrapper.find('span.value');
  expect(span.text()).toEqual('United Kingdom');
});

it('should have the class passed', () => {
  expect(wrapper.first().hasClass('test')).toEqual(true);
});
