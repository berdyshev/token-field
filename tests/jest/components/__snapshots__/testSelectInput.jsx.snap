// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+++capturing Snapshot of SelectInput 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      <span>
        United Kingdom
      </span>
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of SelectInput simple 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      <span>
        3
      </span>
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of SelectInput value 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      <span>
        3
      </span>
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of SelectInput value 2`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      <span>
        large
      </span>
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of SelectInput with icon 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      <span>
        <i
          className="fa fa-thermometer-three-quarters dp-icon dp-icon--s"
        />
        <span>
          Â 
        </span>
        Warm
      </span>
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;
